lib/web_ui/lib/src/engine/semantics/dialog.dart

```diff
@@ -13,26 +13,84 @@ class Dialog extends RoleManager {
   Dialog(SemanticsObject semanticsObject) : super(Role.dialog, semanticsObject);
 
   @override
-  void dispose() {
-    semanticsObject.element.removeAttribute('aria-label');
-    semanticsObject.clearAriaRole();
+  void update() {
+    // If semantic object corresponding to the dialog also provides the label
+    // for itself it is applied as `aria-label`. See also [describeBy].
+    if (semanticsObject.namesRoute) {
+      final String? label = semanticsObject.label;
+      assert(() {
+        if (label == null || label.trim().isEmpty) {
+          printWarning(
+            'Semantic node ${semanticsObject.id} had both scopesRoute and '
+            'namesRoute set, indicating a self-labelled dialog, but it is '
+            'missing the label. A dialog should be labelled either by setting '
+            'namesRoute on itself and providing a label, or by containing a '
+            'child node with namesRoute that can describe it with its content.'
+          );
+        }
+        return true;
+      }());
+      semanticsObject.element.setAttribute('aria-label', label ?? '');
```

Comment:
I don't see where the `aria-label` is set for the case where `namesRoute` is a descendant of `scopesRoute`.

Does it happen through the `LabelAndValue` role manager? If yes, then why are we doing it in the `Dialog` role manager too?

Reply:
No, we don't set `aria-label` when a descendant provides  a description. We use `aria-describedby="ID_OF_DESCENDANT"` instead.

Reply:
And then that descendant is supposed to set the `aria-label`, right? I don't see `RouteName` doing it, so I'm assuming that `LabelAndValue` is doing it. But `LabelAndValue` should be doing it for `Dialog` as well. Maybe I'm missing something here.

Reply:
Ah I see. I was confused by this test:
```dart
expectSemanticsTree('''
  <sem aria-describedby="flt-semantic-node-2" style="$rootSemanticStyle">
    <sem-c>
      <sem>
        <sem-c>
          <sem aria-label="$label"></sem>
        </sem-c>
      </sem>
    </sem-c>
  </sem>
''');
```

Reply:
Oh, it's because the descendant can provide the description by applying `aria-label` to itself, but it can also be a group of elements that provide a description together.